keps/2349-multikueue-external-custom-job-support/README.md

```diff
@@ -0,0 +1,235 @@
+# KEP-2349: MultiKueue external custom Job support
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+  - [Constraints](#constraints)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Risks](#risks)
+    - [Mitigations](#mitigations)
+- [Design Details](#design-details)
+  - [Generic Adapter via ConfigMap](#generic-adapter-via-configmap)
+    - [Job Flow from Manager to Worker](#job-flow-from-manager-to-worker)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+This KEP introduces a mechanism to support MultiKueue for custom Jobs.  This introduces a generic, built-in MultiKueue adapter that is configurable via the Kueue Config.
+
+## Motivation
+
+Many Kueue adopters have specific CustomResources which might not be public.
+Kueue should give a possibility to manage such custom Jobs across multiple clusters.
+
+### Goals
+
+Give a possibility for external controllers to manage custom Jobs in a MultiKueue environment.
+
+
+### Non-Goals
+
+## Proposal
+
+MultiKueue controller will be configurable via configmap. Based on this configmap, a generic MultiKueueAdapter will be implemented within Kueue. When processing a workload without supported built-in adapter, generic MultiKueueAdapter will be called based on GroupVersionKind (GVK) defined in configmap. It will use configuration defined to manage lifycycle's on the worker clusters.
+
+### User Stories
+
+As a cluster administrator I would like to have the ability to schedule Jobs
+which aren't natively supported by Kueue, eg Tekton.
+
+### Constraints
+
+- Limited Transformation: Generic Adaptor are limited to simple JsonPatch transformation configurable.
+
+### Risks and Mitigations
+
+#### Risks
+
+ Configmaps are unstructured by default. They don't provide validation by default unlike CRD. User could make typo mistakes or give invalid syntax in configmap. Or user could give invalid syntax.
+
+#### Mitigations
+
+Configmap which aren't correct should logged like Invalid GVK format, Invalid JsonPatch ops. We could have cli command to validate adapter logic.
+
+## Design Details
+
+### Generic Adapter via ConfigMap
+
+The `kueue-manager-config` configmap will be expanded to configure the generic adapter.
+
+```
+     # In the multikueue Kueue Config
+     multikueue:
+       externalFrameworks:
```

Comment:
I might be able to approve only this `externalFrameworks` field since the kueue-controller-manager should recognize which frameworks are managed by external-controller.

Is this understanding right?

Reply:
If we introduce this API, I would like to keep the same API surface like

```yaml
multiKueue:
  externalFrameworks: ["PipelineRun.v1.pipelinerun.tekton.dev"]
```

https://github.com/kubernetes-sigs/kueue/blob/b44cf15a52b5c97d46d172dcf2432228bd03403e/apis/config/v1beta1/configuration_types.go#L396-L398

Reply:
I was also thinking of going with this, but the problem is that this doesn't reliably extract gvk unless we use g/v/k format.